File: dictionary/zig/algorithms.zig
const std = @import("std");

// vibetyper Zig dictionary — algorithms

fn binary_search(sorted: []const i32, target: i32) ?usize {
    var lo: usize = 0;
    var hi: usize = sorted.len;
    while (lo < hi) {
        const mid = lo + (hi - lo) / 2;
        if (sorted[mid] == target) return mid;
        if (sorted[mid] < target) {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    return null;
}

fn insertion_sort(values: []i32) void {
    var i: usize = 1;
    while (i < values.len) : (i += 1) {
        const key = values[i];
        var j = i;
        while (j > 0 and values[j - 1] > key) : (j -= 1) {
            values[j] = values[j - 1];
        }
        values[j] = key;
    }
}

fn selection_sort(values: []i32) void {
    var i: usize = 0;
    while (i < values.len) : (i += 1) {
        var best = i;
        var j = i + 1;
        while (j < values.len) : (j += 1) {
            if (values[j] < values[best]) best = j;
        }
        const tmp = values[i];
        values[i] = values[best];
        values[best] = tmp;
    }
}

fn merge_sort_alloc(allocator: std.mem.Allocator, values: []const i32) ![]i32 {
    if (values.len <= 1) return allocator.dupe(i32, values);
    const mid = values.len / 2;
    const left = try merge_sort_alloc(allocator, values[0..mid]);
    defer allocator.free(left);
    const right = try merge_sort_alloc(allocator, values[mid..]);
    defer allocator.free(right);
    const out = try allocator.alloc(i32, values.len);
    var i: usize = 0; var j: usize = 0; var k: usize = 0;
    while (i < left.len and j < right.len) : (k += 1) {
        if (left[i] <= right[j]) { out[k] = left[i]; i += 1; } else { out[k] = right[j]; j += 1; }
    }
    @memcpy(out[k..], left[i..]);
    @memcpy(out[k + left.len - i ..], right[j..]);
    return out;
}

fn quick_sort(values: []i32) void {
    if (values.len <= 1) return;
    const pivot = values[values.len - 1];
    var i: usize = 0;
    var j: usize = 0;
    while (j < values.len - 1) : (j += 1) {
        if (values[j] <= pivot) {
            std.mem.swap(i32, &values[i], &values[j]);
            i += 1;
        }
    }
    std.mem.swap(i32, &values[i], &values[values.len - 1]);
    quick_sort(values[0..i]);
    quick_sort(values[i + 1 ..]);
}

fn counting_sort_alloc(allocator: std.mem.Allocator, values: []const u32, max_value: u32) ![]u32 {
    const counts = try allocator.alloc(usize, max_value + 1);
    defer allocator.free(counts);
    @memset(counts, 0);
    for (values) |v| counts[v] += 1;
    var out = std.ArrayList(u32).init(allocator);
    for (counts, 0..) |c, v| {
        var n = c;
        while (n > 0) : (n -= 1) try out.append(@intCast(v));
    }
    return out.toOwnedSlice();
}

fn bfs_can_reach(allocator: std.mem.Allocator, graph: []const []const usize, start: usize, goal: usize) !bool {
    var queue = std.ArrayList(usize).init(allocator); defer queue.deinit();
    const visited = try allocator.alloc(bool, graph.len); defer allocator.free(visited);
    @memset(visited, false);
    visited[start] = true; try queue.append(start);
    var head: usize = 0;
    while (head < queue.items.len) : (head += 1) {
        const node = queue.items[head];
        if (node == goal) return true;
        for (graph[node]) |next| if (!visited[next]) {
            visited[next] = true;
            try queue.append(next);
        };
    }
    return false;
}

fn dfs_has_cycle(allocator: std.mem.Allocator, graph: []const []const usize) !bool {
    const state = try allocator.alloc(u8, graph.len); defer allocator.free(state);
    @memset(state, 0);
    const Visit = struct {
        fn run(graph: []const []const usize, state: []u8, node: usize) bool {
            state[node] = 1;
            for (graph[node]) |next| {
                if (state[next] == 1) return true; if (state[next] == 0 and run(graph, state, next)) return true;
            }
            state[node] = 2;
            return false;
        }
    };
    for (graph, 0..) |_, root| if (state[root] == 0 and Visit.run(graph, state, root)) return true;
    return false;
}

fn connected_components(allocator: std.mem.Allocator, graph: []const []const usize) !usize {
    const visited = try allocator.alloc(bool, graph.len); defer allocator.free(visited);
    @memset(visited, false);
    const Flood = struct {
        fn fill(graph: []const []const usize, visited: []bool, node: usize) void {
            visited[node] = true;
            for (graph[node]) |next| if (!visited[next]) fill(graph, visited, next);
        }
    };
    var components: usize = 0;
    for (graph, 0..) |_, node| if (!visited[node]) {
        components += 1;
        Flood.fill(graph, visited, node);
    };
    return components;
}

fn dijkstra_shortest(allocator: std.mem.Allocator, graph: []const []const u64, start: usize) ![]u64 {
    const dist = try allocator.alloc(u64, graph.len); const visited = try allocator.alloc(bool, graph.len);
    @memset(dist, std.math.maxInt(u64));
    @memset(visited, false);
    dist[start] = 0;
    var done: usize = 0;
    while (done < graph.len) {
        var best: ?usize = null;
        for (graph, 0..) |_, node| if (!visited[node] and (best == null or dist[node] < dist[best.?])) best = node;
        if (best == null) break;
        visited[best.?] = true;
        done += 1;
        for (graph[best.?], 0..) |weight, next| if (dist[best.?] + weight < dist[next]) dist[next] = dist[best.?] + weight;
    }
    return dist;
}

fn knapsack_01(allocator: std.mem.Allocator, weights: []const u64, values: []const u64, capacity: usize) !u64 {
    const dp = try allocator.alloc([]u64, weights.len + 1);
    defer { for (dp) |row| allocator.free(row); allocator.free(dp); }
    for (dp, 0..) |_, i| dp[i] = try allocator.alloc(u64, capacity + 1);
    @memset(dp[0], 0);
    for (1..dp.len) |i| {
        const w = @as(usize, @intCast(weights[i - 1]));
        for (0..=capacity) |c| {
            dp[i][c] = dp[i - 1][c];
            if (c >= w) dp[i][c] = @max(dp[i][c], dp[i - 1][c - w] + values[i - 1]);
        }
    }
    return dp[weights.len][capacity];
}

fn coin_change_min(allocator: std.mem.Allocator, coins: []const u32, amount: usize) !?usize {
    const dp = try allocator.alloc(?usize, amount + 1);
    defer allocator.free(dp);
    @memset(dp, null);
    dp[0] = 0;
    for (1..=amount) |a| {
        for (coins) |coin| {
            const c = @as(usize, @intCast(coin));
            if (c <= a and dp[a - c] != null) {
                const candidate = dp[a - c].? + 1;
                if (dp[a] == null or candidate < dp[a].?) dp[a] = candidate;
            }
        }
    }
    return dp[amount];
}

fn lcs_length(allocator: std.mem.Allocator, a: []const u8, b: []const u8) !usize {
    const dp = try allocator.alloc([]usize, a.len + 1);
    defer allocator.free(dp);
    for (dp, 0..) |_, i| dp[i] = try allocator.alloc(usize, b.len + 1);
    for (dp) |row| @memset(row, 0);
    for (1..=a.len) |i| {
        for (1..=b.len) |j| {
            dp[i][j] = if (a[i - 1] == b[j - 1]) dp[i - 1][j - 1] + 1 else @max(dp[i - 1][j], dp[i][j - 1]);
        }
    }
    return dp[a.len][b.len];
}

fn lis_length(allocator: std.mem.Allocator, values: []const i32) !usize {
    if (values.len == 0) return 0;
    const dp = try allocator.alloc(usize, values.len);
    defer allocator.free(dp);
    @memset(dp, 1);
    var best: usize = 1;
    for (1..values.len) |i| {
        for (0..i) |j| {
            if (values[j] < values[i]) dp[i] = @max(dp[i], dp[j] + 1);
        }
        best = @max(best, dp[i]);
    }
    return best;
}

fn kadane_max_subarray(values: []const i32) i32 {
    var best = values[0];
    var current = values[0];
    for (values[1..]) |v| {
        current = @max(v, current + v);
        best = @max(best, current);
    }
    return best;
}

fn two_sum_sorted(sorted: []const i32, target: i32) ?struct { i: usize, j: usize } {
    var lo: usize = 0;
    var hi: usize = sorted.len - 1;
    while (lo < hi) {
        const sum = sorted[lo] + sorted[hi];
        if (sum == target) return .{ .i = lo, .j = hi };
        if (sum < target) { lo += 1; } else { hi -= 1; }
    }
    return null;
}

fn pair_with_sum(allocator: std.mem.Allocator, values: []const i32, target: i32) !?struct { i: usize, j: usize } {
    var seen = std.StringHashMap(usize).init(allocator);
    defer seen.deinit();
    var buf: [32]u8 = undefined;
    for (values, 0..) |v, i| {
        const want = try std.fmt.bufPrint(&buf, "{d}", .{target - v});
        if (seen.get(want)) |j| return .{ .i = j, .j = i };
        const key = try std.fmt.bufPrint(&buf, "{d}", .{v});
        try seen.put(key, i);
    }
    return null;
}

fn next_permutation(values: []i32) bool {
    var i = values.len - 1;
    while (i > 0 and values[i - 1] >= values[i]) i -= 1;
    if (i == 0) {
        std.mem.reverse(i32, values);
        return false;
    }
    var j = values.len - 1;
    while (values[j] <= values[i - 1]) j -= 1;
    std.mem.swap(i32, &values[i - 1], &values[j]);
    std.mem.reverse(i32, values[i..]);
    return true;
}

fn rotate_matrix_90(allocator: std.mem.Allocator, matrix: []const []const i32) ![][]i32 {
    const n = matrix.len;
    const out = try allocator.alloc([]i32, n);
    for (out, 0..) |_, i| out[i] = try allocator.alloc(i32, n);
    for (matrix, 0..) |row, r| {
        for (row, 0..) |v, c| {
            out[c][n - 1 - r] = v;
        }
    }
    return out;
}

fn spiral_order_alloc(allocator: std.mem.Allocator, matrix: []const []const i32) !std.ArrayList(i32) {
    var result = std.ArrayList(i32).init(allocator);
    if (matrix.len == 0) return result;
    const rows = matrix.len;
    const cols = matrix[0].len;
    var top: usize = 0; var bottom = rows - 1; var left: usize = 0; var right = cols - 1;
    while (top <= bottom and left <= right) {
        for (left..=right) |c| try result.append(matrix[top][c]);
        for (top + 1..=bottom) |r| try result.append(matrix[r][right]);
        if (top < bottom) for (left..=right) |c| try result.append(matrix[bottom][right + left - c]);
        if (left < right) for (top + 1..bottom) |r| try result.append(matrix[right + top - r][left]);
        top += 1; bottom -= 1; left += 1; right -= 1;
    }
    return result;
}

fn matrix_multiply_alloc(allocator: std.mem.Allocator, a: []const []const f64, b: []const []const f64) ![][]f64 {
    const m = a.len;
    const n = b.len;
    const p = b[0].len;
    const out = try allocator.alloc([]f64, m);
    for (out, 0..) |_, i| out[i] = try allocator.alloc(f64, p);
    for (out) |row| @memset(row, 0);
    for (0..m) |i| {
        for (0..n) |k| {
            const ak = a[i][k];
            for (0..p) |j| out[i][j] += ak * b[k][j];
        }
    }
    return out;
}

fn binary_search_rotated(values: []const i32, target: i32) ?usize {
    var lo: usize = 0;
    var hi: usize = values.len;
    while (lo < hi) {
        const mid = lo + (hi - lo) / 2;
        if (values[mid] == target) return mid;
        if (values[lo] <= values[mid]) {
            if (target >= values[lo] and target < values[mid]) { hi = mid; } else { lo = mid + 1; }
        } else {
            if (target > values[mid] and target <= values[hi - 1]) { lo = mid + 1; } else { hi = mid; }
        }
    }
    return null;
}

fn fibonacci_memoized(n: usize, memo: []?u64) u64 {
    if (memo[n]) |cached| return cached;
    const result = if (n < 2) @as(u64, @intCast(n)) else fibonacci_memoized(n - 1, memo) + fibonacci_memoized(n - 2, memo);
    memo[n] = result;
    return result;
}

fn partition_lomuto(values: []i32, lo: usize, hi: usize) usize {
    const pivot = values[hi];
    var i = lo;
    var j = lo;
    while (j < hi) : (j += 1) {
        if (values[j] < pivot) {
            std.mem.swap(i32, &values[i], &values[j]);
            i += 1;
        }
    }
    std.mem.swap(i32, &values[i], &values[hi]);
    return i;
}

File: dictionary/rust/algorithms.rs
// vibetyper Rust dictionary — algorithms

fn binary_search(sorted: &[i32], target: i32) -> Option<usize> {
    let mut low = 0;
    let mut high = sorted.len();
    while low < high {
        let mid = low + (high - low) / 2;
        if sorted[mid] < target {
            low = mid + 1;
        } else if sorted[mid] > target {
            high = mid;
        } else {
            return Some(mid);
        }
    }
    None
}

// Repeatedly pick the smallest remaining value and swap it into place.
fn selection_sort(items: &mut [i32]) {
    for i in 0..items.len() {
        let mut min_index = i;
        for j in (i + 1)..items.len() {
            if items[j] < items[min_index] {
                min_index = j;
            }
        }
        items.swap(i, min_index);
    }
}

fn insertion_sort(items: &mut [i32]) {
    for i in 1..items.len() {
        let key = items[i];
        let mut j = i;
        while j > 0 && items[j - 1] > key {
            items[j] = items[j - 1];
            j -= 1;
        }
        items[j] = key;
    }
}

fn merge_sort(values: Vec<i32>) -> Vec<i32> {
    if values.len() <= 1 {
        return values;
    }
    let mid = values.len() / 2;
    let mut left = merge_sort(values[..mid].to_vec()).into_iter().peekable();
    let mut right = merge_sort(values[mid..].to_vec()).into_iter().peekable();
    let mut merged = Vec::new();
    while left.peek().is_some() || right.peek().is_some() {
        let take_left = left.peek().zip(right.peek()).map_or(true, |(a, b)| a <= b);
        merged.push((if take_left { left.next() } else { right.next() }).unwrap());
    }
    merged
}

fn quicksort(items: &mut [i32]) {
    if items.len() < 2 { return; }
    let pivot = items[items.len() - 1];
    let mut store = 0;
    for i in 0..items.len() - 1 {
        if items[i] < pivot {
            items.swap(i, store);
            store += 1;
        }
    }
    items.swap(store, items.len() - 1);
    let (left, right) = items.split_at_mut(store);
    quicksort(left);
    quicksort(&mut right[1..]);
}

// Counting sort for small, bounded u32 keys.
fn counting_sort(values: &[u32], max_value: u32) -> Vec<u32> {
    let mut counts = vec![0usize; max_value as usize + 1];
    for &value in values {
        counts[value as usize] += 1;
    }
    let mut out = Vec::with_capacity(values.len());
    for (value, &count) in counts.iter().enumerate() {
        out.extend(std::iter::repeat(value as u32).take(count));
    }
    out
}

fn bfs_can_reach(graph: &[Vec<usize>], start: usize, goal: usize) -> bool {
    let mut seen = vec![false; graph.len()];
    let mut queue = std::collections::VecDeque::new();
    seen[start] = true;
    queue.push_back(start);
    while let Some(node) = queue.pop_front() {
        if node == goal { return true; }
        for &next in &graph[node] {
            if !seen[next] {
                seen[next] = true;
                queue.push_back(next);
            }
        }
    }
    false
}

fn dfs_has_cycle(graph: &[Vec<usize>]) -> bool {
    fn visit(node: usize, graph: &[Vec<usize>], state: &mut [u8]) -> bool {
        state[node] = 1;
        for &next in &graph[node] {
            if state[next] == 1 || (state[next] == 0 && visit(next, graph, state)) {
                return true;
            }
        }
        state[node] = 2;
        false
    }
    let mut state = vec![0u8; graph.len()];
    (0..graph.len()).any(|node| state[node] == 0 && visit(node, graph, &mut state))
}

fn connected_components(graph: &[Vec<usize>]) -> usize {
    let mut seen = vec![false; graph.len()];
    let mut components = 0;
    for start in 0..graph.len() {
        if seen[start] { continue; }
        components += 1;
        let mut stack = vec![start];
        seen[start] = true;
        while let Some(node) = stack.pop() {
            for &next in &graph[node] {
                if !seen[next] { seen[next] = true; stack.push(next); }
            }
        }
    }
    components
}

fn dijkstra_shortest(graph: &[Vec<(usize, u64)>], start: usize) -> Vec<u64> {
    let mut dist = vec![u64::MAX; graph.len()];
    let mut heap = std::collections::BinaryHeap::new();
    dist[start] = 0;
    heap.push(std::cmp::Reverse((0, start)));
    while let Some(std::cmp::Reverse((cost, node))) = heap.pop() {
        if cost > dist[node] { continue; }
        for &(next, weight) in &graph[node] {
            let candidate = cost + weight;
            if candidate < dist[next] { dist[next] = candidate; heap.push(std::cmp::Reverse((candidate, next))); }
        }
    }
    dist
}

// 0/1 knapsack with a one-dimensional DP table over capacity.
fn knapsack_01(weights: &[u64], values: &[u64], capacity: u64) -> u64 {
    let mut best = vec![0u64; capacity as usize + 1];
    for (index, &weight) in weights.iter().enumerate() {
        let value = values[index];
        for cap in (weight..=capacity).rev() {
            let take = best[cap as usize - weight as usize] + value;
            best[cap as usize] = best[cap as usize].max(take);
        }
    }
    best[capacity as usize]
}

fn coin_change_min(coins: &[u64], amount: u64) -> Option<usize> {
    let mut fewest = vec![usize::MAX; amount as usize + 1];
    fewest[0] = 0;
    for sum in 1..=amount as usize {
        for &coin in coins {
            if coin as usize <= sum && fewest[sum - coin as usize] != usize::MAX {
                fewest[sum] = fewest[sum].min(fewest[sum - coin as usize] + 1);
            }
        }
    }
    if fewest[amount as usize] == usize::MAX {
        None
    } else {
        Some(fewest[amount as usize])
    }
}

fn lcs_length(a: &str, b: &str) -> usize {
    let a: Vec<char> = a.chars().collect();
    let b: Vec<char> = b.chars().collect();
    let mut table = vec![vec![0usize; b.len() + 1]; a.len() + 1];
    for i in 0..a.len() {
        for j in 0..b.len() {
            table[i + 1][j + 1] = if a[i] == b[j] {
                table[i][j] + 1
            } else {
                table[i][j + 1].max(table[i + 1][j])
            };
        }
    }
    table[a.len()][b.len()]
}

fn lis_length(values: &[i32]) -> usize {
    let mut tails: Vec<i32> = Vec::new();
    for &value in values {
        let pos = tails.partition_point(|&tail| tail < value);
        if pos == tails.len() {
            tails.push(value);
        } else {
            tails[pos] = value;
        }
    }
    tails.len()
}

// Kadane algorithm: largest sum of any contiguous subarray.
fn kadane_max_subarray(values: &[i32]) -> i32 {
    let mut best = i32::MIN;
    let mut current = 0;
    for &value in values {
        current = current.max(0) + value;
        best = best.max(current);
    }
    best
}

fn two_sum_sorted(sorted: &[i32], target: i32) -> Option<(usize, usize)> {
    let mut left = 0;
    let mut right = sorted.len().saturating_sub(1);
    while left < right {
        let sum = sorted[left] + sorted[right];
        if sum == target {
            return Some((left, right));
        }
        if sum < target {
            left += 1;
        } else {
            right -= 1;
        }
    }
    None
}

// One pass with a hash map recording where each value was seen.
fn pair_with_sum(values: &[i32], target: i32) -> Option<(usize, usize)> {
    use std::collections::HashMap;
    let mut seen = HashMap::new();
    for (index, &value) in values.iter().enumerate() {
        if let Some(&other) = seen.get(&(target - value)) {
            return Some((other, index));
        }
        seen.insert(value, index);
    }
    None
}

fn next_permutation(values: &mut [i32]) -> bool {
    let Some(pivot) = (0..values.len().saturating_sub(1))
        .rev()
        .find(|&i| values[i] < values[i + 1])
    else {
        return false;
    };
    let swap_with = (pivot + 1..values.len())
        .rev()
        .find(|&j| values[j] > values[pivot])
        .unwrap();
    values.swap(pivot, swap_with);
    values[pivot + 1..].reverse();
    true
}

fn rotate_matrix_90(matrix: &mut [Vec<i32>]) {
    let size = matrix.len();
    for layer in 0..size / 2 {
        for offset in 0..size - 2 * layer - 1 {
            let top = matrix[layer][layer + offset];
            matrix[layer][layer + offset] = matrix[size - 1 - layer - offset][layer];
            matrix[size - 1 - layer - offset][layer] = matrix[size - 1 - layer][size - 1 - layer - offset];
            matrix[size - 1 - layer][size - 1 - layer - offset] = matrix[layer + offset][size - 1 - layer];
            matrix[layer + offset][size - 1 - layer] = top;
        }
    }
}

fn spiral_order(matrix: &[Vec<i32>]) -> Vec<i32> {
    if matrix.is_empty() || matrix[0].is_empty() { return Vec::new(); }
    let mut top = 0;
    let mut bottom = matrix.len() - 1;
    let mut left = 0;
    let mut right = matrix[0].len() - 1;
    let mut out = Vec::new();
    while top <= bottom && left <= right {
        for col in left..=right { out.push(matrix[top][col]); }
        for row in top + 1..=bottom { out.push(matrix[row][right]); }
        if top < bottom { for col in (left..right).rev() { out.push(matrix[bottom][col]); } }
        if left < right { for row in (top + 1..bottom).rev() { out.push(matrix[row][left]); } }
        top += 1; bottom -= 1; left += 1; right -= 1;
    }
    out
}

// Naive row-by-column product; loop order is cache friendly.
fn matrix_multiply(left: &[Vec<f64>], right: &[Vec<f64>]) -> Vec<Vec<f64>> {
    let rows = left.len();
    let cols = right[0].len();
    let inner = right.len();
    let mut out = vec![vec![0.0; cols]; rows];
    for i in 0..rows {
        for k in 0..inner {
            let factor = left[i][k];
            for j in 0..cols {
                out[i][j] += factor * right[k][j];
            }
        }
    }
    out
}

fn binary_search_rotated(values: &[i32], target: i32) -> Option<usize> {
    let mut low = 0;
    let mut high = values.len();
    while low < high {
        let mid = low + (high - low) / 2;
        if values[mid] == target { return Some(mid); }
        if values[low] <= values[mid] {
            if values[low] <= target && target < values[mid] { high = mid; } else { low = mid + 1; }
        } else if values[mid] < target && target <= values[high - 1] { low = mid + 1; } else { high = mid; }
    }
    None
}

fn fibonacci_memoized(n: usize) -> u64 {
    let mut memo: Vec<Option<u64>> = vec![None; n + 1];
    fn fib(index: usize, memo: &mut [Option<u64>]) -> u64 {
        if let Some(value) = memo[index] {
            return value;
        }
        let value = if index < 2 {
            index as u64
        } else {
            fib(index - 1, memo) + fib(index - 2, memo)
        };
        memo[index] = Some(value);
        value
    }
    fib(n, &mut memo)
}

// Lomuto partition; the returned index is the final position of the pivot.
fn partition_lomuto(items: &mut [i32]) -> usize {
    let pivot = items[items.len() - 1];
    let mut store = 0;
    for i in 0..items.len() - 1 {
        if items[i] <= pivot {
            items.swap(i, store);
            store += 1;
        }
    }
    items.swap(store, items.len() - 1);
    store
}
